website/blog/2021-01-05-macromix.md

```diff
@@ -0,0 +1,183 @@
+<text text-anchor="middle" x="126.84" y="-158.3" font-family="Times,serif" font-size="14.00">app_3.0</text>
+</g>
+<!-- a_library_3.0 -->
+<g id="node2" class="node">
+<title>a_library_3.0</title>
+<ellipse fill="none" stroke="black" cx="57.84" cy="-90" rx="57.69" ry="18"/>
+<text text-anchor="middle" x="57.84" y="-86.3" font-family="Times,serif" font-size="14.00">a_library_3.0</text>
+</g>
+<!-- app_3.0&#45;&gt;a_library_3.0 -->
+<g id="edge1" class="edge">
+<title>app_3.0&#45;&gt;a_library_3.0</title>
+<path fill="none" stroke="black" d="M111.19,-145.12C102.32,-136.12 91.05,-124.68 81.15,-114.65"/>
+<polygon fill="black" stroke="black" points="83.43,-111.97 73.92,-107.31 78.45,-116.89 83.43,-111.97"/>
+</g>
+<!-- b_library_2.13 -->
+<g id="node4" class="node">
+<title>b_library_2.13</title>
+<ellipse fill="none" stroke="black" cx="196.84" cy="-90" rx="63.09" ry="18"/>
+<text text-anchor="middle" x="196.84" y="-86.3" font-family="Times,serif" font-size="14.00">b_library_2.13</text>
+</g>
+<!-- app_3.0&#45;&gt;b_library_2.13 -->
+<g id="edge3" class="edge">
+<title>app_3.0&#45;&gt;b_library_2.13</title>
+<path fill="none" stroke="black" d="M142.38,-145.46C151.46,-136.39 163.09,-124.75 173.28,-114.57"/>
+<polygon fill="black" stroke="black" points="175.78,-117.01 180.38,-107.47 170.83,-112.06 175.78,-117.01"/>
+</g>
+<!-- munit_3.0 -->
+<g id="node3" class="node">
+<title>munit_3.0</title>
+<ellipse fill="none" stroke="black" cx="126.84" cy="-18" rx="47.39" ry="18"/>
+<text text-anchor="middle" x="126.84" y="-14.3" font-family="Times,serif" font-size="14.00">munit_3.0</text>
+</g>
+<!-- a_library_3.0&#45;&gt;munit_3.0 -->
+<g id="edge2" class="edge">
+<title>a_library_3.0&#45;&gt;munit_3.0</title>
+<path fill="none" stroke="black" d="M74.2,-72.41C83.09,-63.39 94.23,-52.09 103.97,-42.21"/>
+<polygon fill="black" stroke="black" points="106.56,-44.57 111.09,-34.99 101.57,-39.65 106.56,-44.57"/>
+</g>
+<!-- b_library_2.13&#45;&gt;munit_3.0 -->
+<g id="edge4" class="edge">
+<title>b_library_2.13&#45;&gt;munit_3.0</title>
+<path fill="none" stroke="black" d="M180.26,-72.41C171.23,-63.39 159.93,-52.09 150.05,-42.21"/>
+<polygon fill="black" stroke="black" points="152.38,-39.58 142.83,-34.99 147.43,-44.53 152.38,-39.58"/>
+</g>
+</g>
+</svg>
+</div>
+<!-- prettier-ignore-end -->
+
+This change is possible thanks to the new `-Ytasty-reader` flag in the Scala
+2.13.4 compiler that enables the Scala 2 compiler to read Scala 3 libraries.
+
+This feature is new and has some limitations that's good to be aware of:
+
+- Scala 2.13.4 can only understand libraries that are published with the old
+  Scala 3.0.0-M1 version (latest is 3.0.0-M3 at the time of this writing). The
+  upcoming Scala 2.13.5 release will be able to understand newer Scala 3.x
+  releases.
+- Some common features in Scala 2 macros don't work in Scala 3. Most notably,
+  you can't use quasiquotes or macro bundles. In the case of MUnit, we had to
```

Comment:
I was confused to find that I could compile a macro bundle using Scala 3.0.0-M1 but the macro expansion failed from Scala 2.13.4, despite the macro bundle having the same bytecode when compiled by either 2.13.4 or 3.0.0-M1, and the tasty to call the macro was the same either way. Something to do with reflective invocation not having the required number of arguments.

It's probably worth opening an issue with the trouble you found, because I don't see at this moment why it should fail.

my example was the following:
```scala
class TpeTagImpl(val c: Context) {
  import c.universe._

  // factory for `final case class TpeTag[T](repr: String)`
  // invoked by `implicit def mkTpeTag[T]: TpeTag[T] = macro TpeTagImpl.mkTag[T]`
  def mkTag[T](implicit T: c.WeakTypeTag[T]): Tree = {
    val tpeString = Literal(Constant(T.tpe.toString))
    // new testframework.TpeTag[T](T.tpe.toString)
    New(appliedType(c.mirror.staticClass(classOf[testframework.TpeTag[_]].getName()).toType, T.tpe), tpeString)
  }

}
```

Reply:
I removed the macro bundle mention. I didn't try using bundles, I just assumed they wouldn't work because I thought they were implemented with macroparadise.

Reply:
I confused bundles with https://github.com/milessabin/macro-compat, which is implemented as a macro annotation to add bundle support in Scala 2.11

Reply:
This is very nice work, I will open the issue then for my example.